service/controller/resource/instance/internal/vmsscheck/guard_job.go

```diff
@@ -13,13 +55,93 @@ func (gj *guardJob) ID() string {
 }
 
 func (gj *guardJob) Run() error {
-	// TODO: Implement VMSS instance check here in following PR.
+	// Still not the time to run the check
+	if !time.Now().After(gj.nextExecutionTime) {
+		return nil
+	}
+
+	var err error
+	gj.allInstancesSucceeded, err = gj.reimageFailedInstances(gj.context, gj.resourceGroup, gj.vmss)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	gj.nextExecutionTime = time.Now().Add(10 * time.Second)
 	return nil
 }
 
 func (gj *guardJob) Finished() bool {
-	// TODO: If any of the VMSS instances are in Failed state, return false here.
-	gj.onFinished()
+	// If any of the VMSS instances are in Failed state, return false here.
+	if !gj.allInstancesSucceeded {
+		return false
+	}
 
+	gj.onFinished()
 	return true
 }
+
+// If any of the instances is not Succeeded, returns false.
+// It reimages instances that are in "Failed" state.
+func (gj *guardJob) reimageFailedInstances(ctx context.Context, rg string, vmssName string) (bool, error) {
+	c, err := getVMsClient(ctx)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	// Get a list of instances in the VMSS.
+	iterator, err := c.ListComplete(ctx, rg, vmssName, "", "", "")
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	// Check for rate limit. If current remaining API calls are less than the desired threshold, we don't proceed.
+	rl3m, rl30m := rateLimitThresholdsFromResponse(iterator.Response().Response)
+	if rl3m < remainingCallsThreshold3m || rl30m < remainingCallsThreshold30m {
+		gj.logger.LogCtx(ctx, "level", "warmomg", "message", fmt.Sprintf("The VMSS API remaining calls are not safe to continue (3m %d/%d, 30m %d/%d)", rl3m, remainingCallsMax3m, rl30m, remainingCallsMax30m))
+		return false, nil
+	}
+
+	allSucceeded := true
+
+	for iterator.NotDone() {
+		instance := iterator.Value()
+
+		gj.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Instance %s has state %s", *instance.Name, *instance.ProvisioningState))
+
+		switch *instance.ProvisioningState {
+		case provisioningStateFailed:
+			// Reimage the instance.
+			gj.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Reimaging instance %s", *instance.Name))
+
+			retries := 3
+			for retries > 0 {
+				_, err := c.Reimage(ctx, rg, vmssName, *instance.InstanceID, nil)
+				if err != nil {
+					gj.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("Error reimaging instance %s: %s", *instance.Name, err.Error()))
+					if retries == 0 {
+						return false, microerror.Mask(err)
+					}
+
+					retries = retries - 1
```

Comment:
This was giving errors to @fiunchinho right?

Reply:
Yes, I fixed it deleting the `break` below. I believe it's not necessary?